File: code/1049.last-stone-weight-ii.go
/*
 * @lc app=leetcode id=1049 lang=golang
 *
 * [1049] Last Stone Weight II
 */

// @lc code=start
func lastStoneWeightII(stones []int) int {
	sum := 0
	for _, stone := range stones {
		sum += stone
	}

	space := sum / 2
	dp := make([]int, space+1)

	for _, stone := range stones {
		for j := space; j >= stone; j-- {
			dp[j] = max(dp[j], dp[j-stone]+stone)
		}
	}

	return sum - 2*dp[space]
}

// @lc code=end

File: code/1049.last-stone-weight-ii.java
/*
 * @lc app=leetcode id=1049 lang=java
 *
 * [1049] Last Stone Weight II
 */

// @lc code=start
class Solution {
    public int lastStoneWeightII(int[] stones) {
        int sum = 0;
        for (int num : stones) {
            sum += num;
        }
        int space = sum / 2;
        int[] dp = new int[space + 1];

        for (int i = 0; i < stones.length; i++) {
            for (int j = space; j >= stones[i]; j--) {
                dp[j] = Math.max(dp[j], dp[j - stones[i]] + stones[i]);
            }
        }
        return sum - 2 * dp[space];
    }
}
// @lc code=end

